File: 20160917/C.py
import sys
import math

tmp = 10**100 - 1

for inp in sys.stdin:
	if (inp == ''):
		continue
	x, p, q = map(int, inp.split())
	if (x == 1):
		print('NaN')
		continue

	tmp2 = int(math.floor(math.log(tmp, x)))

	if (p > tmp2 or q > tmp2):
		print('NaN')
		continue

	xp = x**p
	xq = x**q
	if ((xp - 1) % (xq - 1) != 0):
		print('NaN')
	else:
		print((xp - 1) // (xq - 1))

File: 20160917/C.java
import java.util.*;
import java.math.*;

class C {
	public static void main(String args[]) {
		Scanner in = new Scanner(System.in);
		BigInteger x, xp, xq, tmp, tt;
		int p, q;
		tmp = BigInteger.TEN.pow(100);
		tt = new BigInteger("400");

		while (in.hasNextBigInteger()) {
			x = in.nextBigInteger();
			p = in.nextInt();
			q = in.nextInt();
			if (x.equals(BigInteger.ONE) || p >= 340 || q >= 340) {
				System.out.println("NaN");
				continue;
			}

			xp = x.pow(p);
			xq = x.pow(q);

			if (xp.compareTo(tmp) >= 0 || xq.compareTo(tmp) >= 0) {
				System.out.println("NaN");
			} else {
				xp = xp.subtract(BigInteger.ONE);
				xq = xq.subtract(BigInteger.ONE);
				if (xp.mod(xq).compareTo(BigInteger.ZERO) != 0) {
					System.out.println("NaN");
				} else {
					System.out.println(xp.divide(xq));
				}
			}
		}

		in.close();
	}
}